pyccl/_nonlimber_FKEM.py

```diff
@@ -0,0 +1,273 @@
+    """zero chi_min will result in a divide-by-zero error.
+    If it is zero, we set it to something very small
+    """
+    if chi_min == 0.0:
+        chi_min = 1.0e-6
+    chi_logspace_arr = np.logspace(
+        np.log10(chi_min), np.log10(chi_max), num=Nchi, endpoint=True
+    )
+    dlnr = np.log(chi_max / chi_min) / (Nchi - 1.0)
+    cells = []
+
+    a_arr = ccl.scale_factor_of_chi(cosmo, chi_logspace_arr)
+    growfac_arr = ccl.growth_factor(cosmo, a_arr)
+    avg_a1s = clt1.get_avg_weighted_a()
+    avg_a2s = clt2.get_avg_weighted_a()
+    for el in range(len(ls)):
+        ell = ls[el]
+        cls_nonlimber_lin = 0.0
+        status = 0
+        cl_limber_lin, status = lib.angular_cl_vec_limber(
+            cosmo.cosmo,
+            t1,
+            t2,
+            psp_lin,
+            [ell],
+            integ_types["qag_quad"],
+            1,
+            status,
+        )
+        check(status, cosmo=cosmo)
+        cl_limber_nonlin, status = lib.angular_cl_vec_limber(
+            cosmo.cosmo,
+            t1,
+            t2,
+            psp_nonlin,
+            [ell],
+            integ_types["qag_quad"],
+            1,
+            status,
+        )
+        check(status, cosmo=cosmo)
+
+        """transfer function approximation for the case
+        when it's inseperable in k and a
+        exact for seperable transfer functions
+        """
+
+        # chi-integral integrand splines
+
+        fks_1 = np.zeros((len(kernels_t1), Nchi))
+        transfers_t1 = np.zeros((len(kernels_t1), Nchi))
+        for i in range(len(kernels_t1)):
+            k, fk1 = clt1.get_chi_fft(
+                clt1._trc[i], Nchi, chi_min, chi_max, ell
+            )
+
+            if (k is None) or (fk1 is None):
+                transfer_t1_low = np.array(
+                    clt1.get_transfer(np.log(k_low), a_arr)
+                )
+                transfer_t1_avg = clt1.get_transfer(np.log(k_low), avg_a1s[i])
+
+                fchi1_interp = interp1d(
+                    chis_t1[i], kernels_t1[i], fill_value="extrapolate"
+                )
+                fchi1_arr = (
+                    fchi1_interp(chi_logspace_arr)
+                    * chi_logspace_arr
+                    * growfac_arr
+                    * transfer_t1_low[i]
+                    / transfer_t1_avg[i]
+                )
+                # calls to fftlog to perform integration over chi integrals
+                nu, deriv, plaw = _get_general_params(bessels_t1[i])
+                k, fk1 = _fftlog_transform_general(
+                    chi_logspace_arr,
+                    fchi1_arr.flatten(),
+                    float(ell),
+                    nu,
+                    1,
+                    float(deriv),
+                    float(plaw),
+                )
+                check(status, cosmo=cosmo)
+
+                clt1.set_chi_fft(
+                    clt1._trc[i], Nchi, chi_min, chi_max, ell, k, fk1
+                )
```

Comment:
One last question about the algorithm itself: I fail to see what the advantage is of setting and storing these at the level of the tracers, as opposed to just precomputing them and storing them within this function.

The way I see it (and this may be where my misconception lies), `Nchi`, `chi_min`, `chi_max` and perhaps `ell`, will likely change with each new power spectrum you calculate (since e.g. `chi_min` and `chi_max` actually depend on the specific pair of tracers you're correlating, and not on an individual tracer properties). So you end up storing one of these in the tracer for each individual power spectrum you calculate. If that's the case, you don't really reuse these, so there would be no advantage to precomputing them locally within the function.

A parallel worry of storing these within the tracers is that, if what I'm saying above is correct, the associated tracer-level dictionaries may end up being huge if there's a lot of cross-correlations (if you have N tracers, the number of dictionary elements added up across all tracers would grow like N^3), which could cause memory issues.

Might be that I'm missing something, so let me know what you think.